core/src/dht/dht_reducers.rs

```diff
@@ -0,0 +1,144 @@
+        None => old_store,
+        Some(new_store) => Arc::new(new_store),
+    }
+}
+
+/// Maps incoming action to the correct reducer
+fn resolve_reducer<CAS, EAVS>(action_wrapper: &ActionWrapper) -> Option<DhtReducer<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    match action_wrapper.action() {
+        Action::Commit(_, _) => Some(reduce_commit_entry),
+        Action::GetEntry(_) => Some(reduce_get_entry),
+        Action::AddLink(_) => Some(reduce_add_link),
+        Action::GetLinks(_) => Some(reduce_get_links),
+        _ => None,
+    }
+}
+
+//
+pub(crate) fn reduce_commit_entry<CAS, EAVS>(
+    _context: Arc<Context>,
+    old_store: &DhtStore<CAS, EAVS>,
+    action_wrapper: &ActionWrapper,
+) -> Option<DhtStore<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    let action = action_wrapper.action();
+    let (_entry_type, entry) = match action {
+        Action::Commit(entry_type, entry) => (entry_type, entry),
+        _ => unreachable!(),
+    };
+    // Look in local storage if it already has it
+    if old_store
+        .content_storage()
+        .contains(&entry.address())
+        .unwrap()
+    {
+        // Maybe panic as this should never happen?
+        return None;
+    }
+    // Otherwise add it local storage...
+    let mut new_store = (*old_store).clone();
+    let res = new_store.content_storage_mut().add(entry);
+    if res.is_err() {
+        return None;
+    }
+    // ...and publish to the network
+    new_store.network_mut().publish(entry);
+    Some(new_store)
+}
+
+//
+pub(crate) fn reduce_get_entry<CAS, EAVS>(
+    _context: Arc<Context>,
+    old_store: &DhtStore<CAS, EAVS>,
+    action_wrapper: &ActionWrapper,
+) -> Option<DhtStore<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
+    EAVS: EntityAttributeValueStorage + Sized + Clone + PartialEq,
+{
+    // Get Action's input data
+    let action = action_wrapper.action();
+    let address = unwrap_to!(action => Action::GetEntry);
+    // Look in local storage if it already has it
+    if old_store.content_storage().contains(address).unwrap() {
+        return None;
+    }
+    // Otherwise retrieve it from the network...
+    let entry = Entry::from_content(&old_store.network().clone().get(address));
+    let mut new_store = (*old_store).clone();
+    // ...and add it to the local storage
+    let res = new_store.content_storage_mut().add(&entry);
+    match res {
+        Err(_) => None,
+        Ok(()) => Some(new_store),
+    }
+}
+
+//
+pub(crate) fn reduce_add_link<CAS, EAVS>(
+    _context: Arc<Context>,
+    _old_store: &DhtStore<CAS, EAVS>,
+    _action_wrapper: &ActionWrapper,
+) -> Option<DhtStore<CAS, EAVS>>
+where
+    CAS: ContentAddressableStorage + Sized + Clone + PartialEq,
```

Comment:
seems like a good candidate for a type alias

Reply:
I tried:
```Rust
pub type ConcreteCas = ContentAddressableStorage + Sized + Clone + PartialEq;
```
but it gives me `type aliases cannot be used for traits`

And I tried this:
```Rust
pub trait ConcreteCas: ContentAddressableStorage + Sized + Clone + PartialEq {}
impl ConcreteCas for ContentAddressableStorage {}
impl ConcreteCas for Sized {}
impl ConcreteCas for Clone {}
impl ConcreteCas for PartialEq {}
```
But it also gives me lots of errors.
So can we leave that as a todo for later?